File: app/src/main/java/com/software/jgodort/marvelheroes/dependencyInjection/ApplicationModule.java
package com.software.jgodort.marvelheroes.dependencyInjection;

import android.app.Application;
import android.content.Context;

import com.software.jgodort.marvelheroes.MarvelSampleApp;
import com.software.jgodort.marvelheroes.network.MarvelSampleRepository;
import com.software.jgodort.marvelheroes.network.MarvelSampleRepositoryImpl;
import com.software.jgodort.marvelheroes.threading.Executor;
import com.software.jgodort.marvelheroes.threading.MainThread;
import com.software.jgodort.marvelheroes.threading.impl.MainThreadImpl;
import com.software.jgodort.marvelheroes.threading.impl.ThreadExecutorImpl;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Dagger Module to provides all related with android framework.
 * Created by javie on 07/08/2017.
 */

@Module
public class ApplicationModule {

    private MarvelSampleApp mMarvelSampleApp;

    public ApplicationModule(MarvelSampleApp app){
        mMarvelSampleApp=app;
    }

    @Provides
    @Singleton
    Application providesApplication(){
        return mMarvelSampleApp;
    }

    @Provides
    @Singleton
    Context providesContext(){
        return mMarvelSampleApp.getApplicationContext();
    }

    @Provides
    @Singleton
    Executor providesThreadExecutor(){
        return ThreadExecutorImpl.getInstance();
    }

    @Provides
    @Singleton
    MainThread providesMainThread(){
        return MainThreadImpl.getInstance();
    }

    @Provides
    @Singleton
    MarvelSampleRepository providesMarvelSampleRepository(){
        return new MarvelSampleRepositoryImpl();
    }

}

File: app/src/main/java/com/software/jgodort/marvelheroes/dependencyInjection/module/ApplicationModule.kt
package com.software.jgodort.marvelheroes.dependencyInjection.module

import android.app.Application
import android.content.Context
import com.software.jgodort.marvelheroes.BuildConfig
import com.software.jgodort.marvelheroes.dependencyInjection.scopes.PerApplication
import com.software.jgodort.marvelheroes.domain.MarvelRepository
import com.software.jgodort.marvelheroes.domain.repository.MarvelSampleRepository
import com.software.jgodort.marvelheroes.network.MarvelServiceFactory
import com.software.jgodort.marvelheroes.network.mapper.SuperheroRemoteMapper
import com.software.jgodort.marvelheroes.threading.PostExecutionThread
import com.software.jgodort.marvelheroes.threading.ThreadExecutor
import com.software.jgodort.marvelheroes.threading.impl.JobExecutor
import com.software.jgodort.marvelheroes.threading.impl.UIThread
import dagger.Module
import dagger.Provides

/**
 * Dagger Module to provides all related with android framework.
 */

@Module
open class ApplicationModule() {

  @Provides
  @PerApplication
  fun provideContext(application: Application): Context {
    return application
  }


  @Provides
  @PerApplication
  internal fun provideThreadExecutor(jobExecutor: JobExecutor): ThreadExecutor {
    return jobExecutor
  }

  @Provides
  @PerApplication
  internal fun providePostExecutionThread(uiThread: UIThread): PostExecutionThread {
    return uiThread
  }

  @Provides
  @PerApplication
  internal fun provideMarvelRepository(mapper: SuperheroRemoteMapper): MarvelRepository {
    return MarvelSampleRepository(MarvelServiceFactory.makeMarvelAPI(BuildConfig.DEBUG), mapper)
  }

}